<?xml version="1.0"?> 
<gazebo version="1.0">
  <world name="default">
    <scene>
      <ambient rgba="1.0 1.0 1.0 1"/>
      <background rgba="0 0 0 1.0"/>
      <shadows enabled="true"/>
    </scene>

    <physics type="ode">
      <gravity xyz="0 0 -10.0"/>
      <ode>
        <solver type="quick" dt="0.001" iters="10" sor="1.3"/>
        <constraints cfm="0.0" erp="0.2" contact_max_correcting_vel="100.0" contact_surface_layer="0.0"/>
      </ode>
    </physics>

    <include filename="ground_plane.model"/>

    <!-- 25.3kg total mass -->
    <model name="test_model">
      <origin pose="0 0 0.7 0 0 0"/>
      <link name="link">
        <origin pose="0 0 0 0 0 0"/>
        <inertial mass="1.0">
          <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
        </inertial>

        <collision name="collision">
          <origin pose="0 0 -0.55 0 0 0"/>
          <geometry>
            <box size="1 1 0.1"/>
          </geometry>
          <surface>
            <friction>
              <ode mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <origin pose="0 0 -0.55 0 0 0"/>
          <geometry>
            <box size="1 1 0.1"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>

        <collision name="collision2">
          <origin pose="0 0 0 0 0 0"/>
          <geometry>
            <box size="0.5 0.5 0.5"/>
          </geometry>
          <surface>
            <friction>
              <ode mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual2">
          <origin pose="0 0 0 0 0 0"/>
          <geometry>
            <box size="0.5 0.5 0.5"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>

      <link name="bar_1">
        <origin pose="0.45 0 -0.5 0 0 0"/>
        <inertial mass="0.1">
          <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
        </inertial>
        <collision name="collision">
          <geometry>
            <box size="0.02 1.0 0.02"/>
          </geometry>
          <surface>
            <friction>
              <ode mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <geometry>
            <box size="0.02 1.0 0.02"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="bar_12_joint" type="revolute">
        <parent link="bar_2"/>
        <child link="bar_1"/>
        <origin pose="0 0.5 0 0 0 0"/>
        <axis xyz="0 0 1"/>
      </joint>
      <joint name="bar_13_joint" type="revolute">
        <parent link="bar_3"/>
        <child link="bar_1"/>
        <origin pose="0 -0.5 0 0 0 0"/>
        <axis xyz="0 0 1"/>
      </joint>

      <link name="bar_2">
        <origin pose="0.5 0.5 -0.5 0 0 0"/>
        <inertial mass="0.1">
          <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
        </inertial>
        <collision name="collision">
          <origin pose="-0.025 0 0 0 0 0"/>
          <geometry>
            <box size="0.05 0.02 0.02"/>
          </geometry>
          <surface>
            <friction>
              <ode mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <origin pose="-0.025 0 0 0 0 0"/>
          <geometry>
            <box size="0.05 0.02 0.02"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="bar_2_joint" type="revolute">
        <parent link="link"/>
        <child link="bar_2"/>
        <origin pose="0 0 0 0 0 0"/>
        <axis xyz="0 0 1"/>
      </joint>
      <link name="bar_3">
        <origin pose="0.5 -0.5 -0.5 0 0 0"/>
        <inertial mass="0.1">
          <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
        </inertial>
        <collision name="collision">
          <origin pose="-0.025 0 0 0 0 0"/>
          <geometry>
            <box size="0.05 0.02 0.02"/>
          </geometry>
          <surface>
            <friction>
              <ode mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <origin pose="-0.025 0 0 0 0 0"/>
          <geometry>
            <box size="0.05 0.02 0.02"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="bar_3_joint" type="revolute">
        <parent link="link"/>
        <child link="bar_3"/>
        <origin pose="0 0 0 0 0 0"/>
        <axis xyz="0 0 1"/>
      </joint>


      <link name="wheel_1">
        <origin pose="0.5 0.55 -0.5 1.57079633 0 0"/>
        <inertial mass="1.0">
          <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
        </inertial>

        <collision name="collision">
          <geometry>
            <cylinder length="0.1" radius="0.2"/>
          </geometry>
          <surface>
            <friction>
              <ode fdir="1 0 0" mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <geometry>
            <cylinder length="0.1" radius="0.2"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="wheel_1_joint" type="revolute">
        <parent link="bar_2"/>
        <child link="wheel_1"/>
        <origin pose="0 0 0 0 0 0"/>
        <axis xyz="0 1 0"/>
      </joint>

      <link name="wheel_2">
        <origin pose="0.5 -0.55 -0.5 1.57079633 0 0"/>
        <inertial mass="1.0">
          <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
        </inertial>

        <collision name="collision">
          <geometry>
            <cylinder length="0.1" radius="0.2"/>
          </geometry>
          <surface>
            <friction>
              <ode fdir="1 0 0" mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <geometry>
            <cylinder length="0.1" radius="0.2"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="wheel_2_joint" type="revolute">
        <parent link="bar_3"/>
        <child link="wheel_2"/>
        <origin pose="0 0 0 0 0 0"/>
        <axis xyz="0 1 0"/>
      </joint>

      <link name="wheel_3">
        <origin pose="-0.5 0.55 -0.5 1.57079633 0 0"/>
        <inertial mass="1.0">
          <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
        </inertial>

        <collision name="collision">
          <geometry>
            <cylinder length="0.1" radius="0.2"/>
          </geometry>
          <surface>
            <friction>
              <ode fdir="1 0 0" mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <geometry>
            <cylinder length="0.1" radius="0.2"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="wheel_3_joint" type="revolute">
        <parent link="link"/>
        <child link="wheel_3"/>
        <origin pose="0 0 0 0 0 0"/>
        <axis xyz="0 1 0">
          <dynamics damping="0.0" friction="0.0"/>
        </axis>
      </joint>

      <link name="wheel_4">
        <origin pose="-0.5 -0.55 -0.5 1.57079633 0 0"/>
        <inertial mass="1.0">
          <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
        </inertial>

        <collision name="collision">
          <geometry>
            <cylinder length="0.1" radius="0.2"/>
          </geometry>
          <surface>
            <friction>
              <ode fdir="1 0 0" mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="100"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <geometry>
            <cylinder length="0.1" radius="0.2"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="wheel_4_joint" type="revolute">
        <parent link="link"/>
        <child link="wheel_4"/>
        <origin pose="0 0 0 0 0 0"/>
        <axis xyz="0 1 0">
          <dynamics damping="0.0" friction="0.0"/>
        </axis>
      </joint>

      <link name="sphere_link">
        <origin pose="0 0 0.5 0 0 0"/>
        <inertial mass="10.0">
          <inertia ixx="0.1" ixy="0" ixz="0" iyy="0.1" iyz="0" izz="0.1"/>
        </inertial>
        <collision name="collision">
          <geometry>
            <sphere radius="0.25"/>
          </geometry>
          <surface>
            <friction>
              <ode fdir="1 0 0" mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="10"/>
            </contact>
          </surface>
        </collision>
        <visual name="visual">
          <geometry>
            <sphere radius="0.25"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="joint_sphere" type="revolute">
        <parent link="link"/>
        <child link="sphere_link"/>
        <origin pose="0 0 0 0 0 0"/>
        <axis xyz="0 0 1"/>
      </joint>

      <link name="cylinder_link">
        <origin pose="0 0 0.25 0 0 0"/>
        <inertial mass="10.0">
          <inertia ixx="0.1" ixy="0" ixz="0" iyy="0.1" iyz="0" izz="0.1"/>
        </inertial>

        <collision name="collision">
          <geometry>
            <cylinder radius="0.25" length="0.5"/>
          </geometry>
          <surface>
            <friction>
              <ode mu="1" mu2="1"/>
            </friction>
            <contact>
              <ode max_vel="10"/>
            </contact>
          </surface>
        </collision>

        <visual name="visual">
          <geometry>
            <cylinder radius="0.25" length="0.5"/>
          </geometry>
          <material script="Gazebo/WoodPallet"/>
        </visual>
      </link>
      <joint name="joint_cylinder" type="revolute">
        <parent link="sphere_link"/>
        <child link="cylinder_link"/>
        <origin pose="0 0 0 0 0 0"/>
        <axis xyz="0 0 1"/>
      </joint>
      <plugin name="joint_test_plugin" filename="libJointTestPlugin.so">
        <steer>bar_2_joint</steer>
        <steer_pid>10 0.1 0.5</steer_pid>
        <steer_ilim>-10 10</steer_ilim>
        <steer_pos>0.0</steer_pos> <!-- steer -->
        <steer_vel>0</steer_vel> <!-- not used -->
        <steer_eff>100.0</steer_eff> <!-- max pos pid effort -->
        <right>wheel_3_joint</right>
        <right_pid>0.1 0.0 0.001</right_pid>
        <right_ilim>-.1 .1</right_ilim>
        <right_pos>0</right_pos> <!-- final position pid -->
        <right_vel>0</right_vel> <!-- final velocity pid -->
        <right_eff>0.1</right_eff> <!-- drive torque -->
        <left>wheel_4_joint</left>
        <left_pid>0.1 0.0 0.001</left_pid>
        <left_ilim>-.1 .1</left_ilim>
        <left_pos>0</left_pos> <!-- final position pid -->
        <left_vel>0</left_vel> <!-- final velocity pid -->
        <left_eff>0.1</left_eff> <!-- drive torque -->
      </plugin>
    </model>

    <light type="directional" name="sun" cast_shadows="true">
      <origin pose="0 0 10 0 0 0"/>
      <diffuse rgba=".8 .8 .8 1"/>
      <specular rgba=".1 .1 .1 1"/>
      <attenuation range="10" constant="0.8" linear="0.01" quadratic="0.0"/>
      <direction xyz="0 .5 -.5"/>
    </light>

  </world>
</gazebo>

